Sorcery tests

GarOfMeridian opened the pull request:
Sorcery is a school designed to be useful to a huge variety of character types and situations. It has very strong themes of metamagic abilities - changing how spells work or how they are cast for the Sorcerer. The abilities aren't necessarily _better_, just different.

The school has also been designed with "optioning" in mind; as in, there are clear packages that players can take to benefit specific types of characters. For example, 5 Intellect investment gets a fighter (among other things) Charged Clutch, Efficiency, Blood Magic (since they don't have much mana) and (most notably) Devotion, which will help with holy and unholy dedicated weapons in PvP. 

Level 3 is hybrid and/or Qor focused, bringing us some cold and acid damage masteries, stronger reservations, and a mana-intensive ability to help casting when it is really needed; and level 4 is bit more magey, focusing on fire and shock and a much larger resource pool for long fights. Levels 5 and 6 are just plain awesome in their own right and change how a character is played now that they're more Sorcerer than not.

Ideally, a character with levels in Sorcery is one finding a variety of benefits regardless of build. There are purposeful tensions all throughout the school: do I reserve more buffs, thus restricting my mana? I might do that as a fighter. Do I focus on spells to take advantage of all the casting benefit skills? Or do I go halfway to benefit a little from everything, especially damage masteries? It's all up to you. Maybe my favorite little tension is the fact that Crystalize Mana (Jala 6) is weaker the more buffs you reserve, thus killing any "super optimum" mage build and likely splitting mage archetypes into two different paths (huge mana pools or Sorcerer abilities, but not both).

---

Final skill list looks like:
**Level 1**
_Efficiency_ - generalized vigor cost reduction for spells, extremely slow to improve, max 20%
_Minor Spirit Link_ - reserve level 1 buffs
_Charged Clutch_ - empower weapons with your active touch spell for mana per swing, min 4 mana

**Level 2**
_Devotion_ - damage mastery for holy/unholy, max 2 bonus damage (no resistance interaction)
_Moderate Spirit Link_ - reserve level 2 buffs
_Blood Magic_ - use health if out of reagents (min cost in health is the spell's mana cost)

**Level 3**
_Entropy_ - damage mastery for cold/acid, max 2 bonus damage (no resistance interaction)
_Spirit Link_ - reserve level 3 buffs
_Force Spike_ - use extra mana to cast anyway despite a fizzle (min cost is spell's mana cost again)

**Level 4**
_Energy_ - damage mastery for fire/shock, max 2 bonus damage (no resistance interaction)
_Major Spirit Link_ - reserve level 4 buffs
_Will Surge_ - use vigor if out of mana (no synergy with Efficiency) (min cost is spell's mana cost)

**Level 5**
_Master Spirit Link_ - reserve level 5 buffs
_Volatile Focus_ - cast immediately if trance disrupted, unable to attack or cast for duration of time penalty

**Level 6**
??
Currently working out the consequences of these suggestions, of which I would like to choose one or two:

_Finesse_ - this master Sorcerer skill has a chance, based on %, to extend the range of any touch spell or blast spell by 1.

_Spirit Echo_ - when Purged, reserved buffs are only temporarily suppressed, returning after a short time.

_Somancy_ - the master Sorcerer can cast targetless spells (including their own now self-only reserved buffs) while held, blind, or dazzled. Spellpower is cut by at least half.

_Empowerment_ - a master Sorcerer can permanently choose one element to empower. All attacks done by that element, whether weapon or spell, penetrate up to 20% resistance (50% resist becomes 30% for the purpose of the spell, 20% becomes 0, 0 becomes -20%, etc). Additionally, attack spells of that element can't be Deflected.

---

**Level 1: Efficiency**

  "Initiates of Sorcery first take on the arduous task of practicing "
      "spellcasting over and over until it becomes almost automatic. This "
      "endless journey often lasts until long after the initiate has become "
      "a master."

Has a chance to reduce vigor costs of spells up to 20% at 99. Improves extremely slowly.

This spell rounds out level 1 with a skill that is useful to every character type, and which is "weak", but also subtly powerful and useful.

---

**Mana reservation**
This pull includes the five mana reservation skills for testing purposes.

If you have, for example, minor spirit link, level 1 personal enchantments you cast become self-cast only, reserve a percentage of your maximum mana, and never expire. The percentage of your max mana reserved depends on your % in the link skill and your Mysticism stat (to a much smaller degree). Each spell has its own base reservation amount. Incredibly useful spells like Invisibility or Detect Invisibility will obviously have higher reservations than spells like Night Vision.

Example:
Gar buys the level 3 spirit link. He casts Night Vision. It automatically targets himself (as it is now self-cast only), never expires, and his max mana drops by 10%. If he works level 3 spirit link to 99%, his reservation cost for Night Vision will drop to 5%. At the same time, Magic Shield starts out reserving 20% of his mana, and drops to 10% when he maxes spirit link. If he has 128 mana, Night Vision would end up costing 7 max mana, and Magic Shield 14. If he kept both up, his max mana would be 107.

Any potions he uses for buffs work normally (i.e. don't reserve mana).

Currently there are five skills, one for each level up to five. Gaze of the Basilisk, being the only level 6 personal enchantment, cannot be reserved.

Reserved spells can't be passively improved.

Reductions due to reservation do **not** hinder mana regeneration. If you normally have 128 max mana, and you reserve down to 30, you will still regenerate at the same rate as if you had 128 max mana. Circlets are also accounted for and do help.

---

**Also included are a variety of sacrifice skills.** They all range from an initial 300% casting cost penalty down to 100% penalty at 99 in both the skill itself and the spell being used.

Charged Clutch, level 1. If you've got a touch spell going, and you use a melee weapon, you have a chance to empower your strike with the element of that touch spell for a mana cost between 4 and 12 depending on skill in charged clutch and the touch spell. Riija Sword deals random damage out of the main four types. Nerudite Sword resists elements and doesn't work with Charged Clutch.

Blood Magic, level 2. If you're out of reagents, sacrifice an amount of health based on casting cost to power the spell anyway. For example, if you don't have two blue dragon scales, you can still cast Armor of Gort. It's just going to cost you at least 15 health, and as much as 45 if you're low % in both Blood Magic and Gort (i.e. working spells this way is not recommended). Blood magic can't kill you - it'll simply fail - but that doesn't mean you won't find yourself at extremely low health and prone to a gruesome death.

Force Spike, level 3. If you fizzle, sacrifice an amount of mana based on casting cost to ensure the spell fires anyway. For example, you go to Blind someone, and your spell would fizzle - instead, you expend as little as 14 more mana (blind's cost) and as much as 42 (if you're terrible) and the spell no longer fizzles.

Will Surge, level 4. If you're out of mana, sacrifice an amount of vigor based on casting cost to ensure the spell fires anyway. For example, if you're out of mana, and you go to cast blind, you expend an extra 14 vigor (minimum) up to 42 (if you're terrible) and the spell fires anyway. Obviously, it's a terrible idea to rely on this ability, but, as with all these skills, it will help your magic remain reliable when you REALLY need it.

Volatile Focus, level 5. If you're focusing on a spell and it gets disrupted by any means, sacrifice an amount of time based on trance time to ensure the spell fires anyway. (Example: you begin casting Earthquake and then run or get hit. The Earthquake fires immediately. You can't attack or cast spells for a number of seconds based on the time remaining in the trance, your skill at Volatile Focus, and your ability in the spell).

Epiphany, level 6. If you're out of reagents, out of mana, and you fizzle an attempt at a focus spell, (i.e. all of your other sacrifice skills activate all at once), you get a free T-shirt that says "I survived reading the Sorcery school pull request."

Note that none of these skills take spellpower into account except where it factors in naturally (mana costs, trance times, and so on). Low spellpower punishes itself in terms of spell effectiveness. These are skills, and therefore useable at any time.

---

**Damage Masteries**

Three skills, in levels 2, 3, and 4, each have a chance, based on %, to
add up to 2 damage to the elements they master.

level 2 - devotion - holy and unholy
level 3 - entropy - cold and acid
level 4 - energy - fire and shock

This is bonus damage that does not get factored into resistances or
weaknesses - it will cut through resists, but will also not be enhanced
by weaknesses.

It also contributes from ANY source - touch spells, enchanted weapons,
walls, bolts, you name it. The order in terms of levels were chosen to
encourage players picking up parts of Sorcery on characters that aren't
pure mages. Holy and Unholy, in level 2 for example, is a great 5 int
option for fighters who want to do 2 more damage with holy and unholy
dedicated weapons.

level 3, cold and acid, could be interesting for qors and players with
level 2 faren (often taken for resists). Because charged clutch cold or acid +
entropy boosts damage.

level 4, fire and shock, clearly for mage heavy characters.

M59Gar committed:
Spirit links

skittles1 commented on kod/object/passive/spell/persench.kod:
```diff
@@ -162,7 +164,16 @@ messages:
-      local oTarget, lType;
+      local oTarget, lType, iLevel;
+      
+      iLevel = Send(self,@GetLevel);
```
Can use viSpell_level here (@GetLevel always just returns this classvar).

skittles1 commented on kod/object/passive/spell/persench.kod:
```diff
@@ -164,6 +167,15 @@ messages:
       local oTarget, lType;
 
+      if NOT bItemCast
+         AND Send(self,@CanBeReserved)
```
And if @CanBeReserved is called on self here (and always just returns the classvar) we can change this line to AND vbCanBeReserved

M59Gar committed:
Blood Magic Ability

Can cast without reagents by taking health damage. Less health with
skill in skill and spell. Minimum health cost is the mana cost of the
spell.

M59Gar committed:
Force Spike

Force Spike allows a mage to expend a huge quantity of mana to power
through a fizzle. Amount expended starts at triple the cost of the
spell, then relies on spellpower, skill in the skill and spell, and
reaches a minimum at 1x the cost of the spell (paid a second time).

M59Gar committed:
Charged Clutch

Sorcery level 1. Allows an active touch spell to empower a wielder's
melee weapon with that element. Costs mana, dependent on skill in
Charged Clutch and the touch spell. Has a chance of working dependent on
skill in Charged Clutch.

M59Gar committed:
Efficiency

Level 1 skill. Very hard to improve. Has a chance to reduce spell vigor
costs up to 20% at 99, minimum 1.

M59Gar committed:
Local Variable name

M59Gar committed:
Charged Clutch damaging_now var

M59Gar committed:
Will surge text change

M59Gar committed:
Volatile focus fix

skittles1 commented:
This is massive and awesome... I have one request that you might not like but I think we should do (and which we've had to go back and do for other things): the code in Player that allows all this to work should probably be in Battler to allow future monsters, bosses or NPCs to access it also (or central code placed in Battler, and specific variants placed in Player/Monster etc.). It's obviously not a huge issue if we don't do that, but I feel like if we don't do it now, we would want to at some point in the future anyway.

Perhaps it might not be worth it if the skills don't change the way monsters behave that much or at all, but if there's anything we could see a future monster using it could be useful.

GarOfMeridian commented:
Seems like a good idea. I even thought about it at the time. The only reason I didn't go that route was because I actually don't see much of this mattering to monsters. They don't use reagents, don't fizzle, don't have vigor, and so on. They won't need damage masteries, either, because we can just adjust their damage.

Reserved buffs could be interesting on monsters, especially since I do want to get monsters into the same buff and casting abilities as players. I'll take a look at moving that up a level.

skittles1 commented:
I like Finesse and Empowerment for the synergy with Faren (or Qor, make it apply to acid too). I do kind of agree there with the monster stuff (especially the effort involved in moving it for something we probably aren't going to use for them).

GarOfMeridian commented:
I wonder if we should do our first "choice" mechanic with level 6. Put in both Finesse and Empowerment, but only let players choose one? I think I personally favor Empowerment because it is not strictly spell-based... it can help any character type, whether through dedicated weapons, touch spells, whatever.

M59Gar committed:
Sorcery School Expansion setting

Can turn on or off the sorcery school as sold by NPCs.

M59Gar committed:
Fix to )

M59lost commented:
This is too much. The name sucks and this school does not fit in.

Daenks commented:
@M59Lost please feel free to share your opinions, but keep insults out of it. "Sucks" "fail" etc. Also please try and offer reasons and justifications. We need solid information and judgments, not whimsical musings.

You think the name is bad. Okay. Why? Too generic? Doesnt fit lore? Have any suggestion? How about Artificery? or Thalmaturgy?

If you're going to comment here, be productive, or don't comment.
